WcaOnRails/app/webpacker/components/RolesTab/RoleForm.jsx

```diff
@@ -23,19 +25,58 @@ const delegateStatusOptions = ['trainee_delegate', 'candidate_delegate', 'delega
 
 export default function RoleForm({ userId, isActiveRole }) {
   const { data, loading, error } = useLoadedData(roleDataUrl(userId, isActiveRole));
+  const regionsFetching = useLoadedData(apiV0Urls.userGroups.list(groupTypes.delegate_regions));
   const { save, saving } = useSaveAction();
   const selectedGroup = groups[0].value;
   const [formValues, setFormValues] = React.useState({});
   const [apiError, setError] = React.useState(false);
   const [finished, setFinished] = React.useState(false);
 
+  const regions = React.useMemo(() => regionsFetching.data?.filter(
+    (group) => !group.parent_group_id,
+  ), [regionsFetching.data]);
+
+  const subRegions = React.useMemo(() => {
+    const subRegionsList = regionsFetching.data?.filter((group) => group.parent_group_id) || [];
+    return Object.groupBy(subRegionsList, (group) => group.parent_group_id);
+  }, [regionsFetching.data]);
+
+  React.useEffect(() => {
+    const loadingCompleted = !loading && !regionsFetching.loading;
+    let regionId = null;
+    let subregionId = null;
+    if (loadingCompleted) {
+      const roleData = data?.roleData || {};
+      if (roleData.regionId && !regions.find((region) => region.id === roleData.regionId)) {
+        // In this case, the regionId is having the subregionId.
+        subregionId = roleData.regionId;
+        regionId = parseInt(Object.keys(subRegions)
+          .find((regionIndex) => subRegions[regionIndex]
+            .find((subregion) => subregion.id === roleData.regionId)), 10);
+      } else {
+        regionId = roleData.regionId;
+      }
+      setFormValues({
+        delegateStatus: delegateStatusOptions[0],
+        location: '',
+        ...(data?.roleData || {}),
+        regionId,
+        subregionId,
+      });
+    }
+  }, [data, loading, regions, regionsFetching.loading, subRegions]);
+
   React.useEffect(() => {
-    setFormValues({
-      delegateStatus: delegateStatusOptions[0],
-      location: '',
-      ...(data?.roleData || {}),
-    });
-  }, [data]);
+    if (formValues.regionId && formValues.subregionId) {
+      const subRegionList = subRegions[formValues.regionId] || [];
+      if (!subRegionList.find((subregion) => subregion.id === formValues.subregionId)) {
```

Comment:
Can you extract the `find` statement inside the if into its own variable and give it a proper name?